enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,192 @@
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendSizeQuota of type *resource.Quantity
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+
+### Implementation Details/Notes/Constraints [optional]
+
+None
+
+### Risks and Mitigations
+
+* We will need to test the larger size to ensure that everything works as expected; especially defragmentation.
+* Defragmentation, and its effects, will need to be tested as the API server will be unresponsive during this time; we will need to understand the general magnitude of defrag time and documentation those expectations.
+* Downgrades may be impacted if:
+  1. The downgrade is to a version without this feature.
+  2. The quota has been set higher than the default.
+  3. The etcd backends are using more space than the default.
+  A mitigation would be to keep the quota at the higher value (with no way of decreasing it), so that the cluster will continue to function and no data will be lost.
+
+### Drawbacks
+
+* There will be a required etcd rollout when changing values, this happens automatically by the Cluster Etcd Operator to re-render the etcd podspec.
+
+## Design Details
+
+### Open Questions [optional]
+
+For each of these, we would need to compare against the current default of 8GiB. Increasing the quota does not necessarily increase the following, but at the extreme case (e.g. increase to 16GiB and using nearly that much) we expect the following to perform differently.
+* How much memory do we require for a given database size?
+* How long does it take to sync a snapshot of a given size to another peer; for reboot/recovery?
+* How long does it take to compact & defrag a database of a given size?
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+We will need to test a cluster with a larger quota to ensure everything works as expected.
+We should also load the database with a larger quota and make note of the memory requirements of a defragmentation.
+  - We should do this with some values we could assume a customer would use; 16GiB, 32GiB, 100GiB (in the extreme).
+We need to test that it is not possible to shrink the quota once it has be set to a larger value.
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+None
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+- Customer feedback
+- API extension
+- Value change with minimal disruption
+- Performance testing for some larger-than-default values
+- End user documentation
+  - Description of the feature and expectations of memory usages for some largers values.
+  - Steps to change the value.
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+N/A
```

Comment:
What happens if they ask for more space than is available, especially in an on-prem configuration where dynamic storage allocation may not be possible?

Reply:
That's a good point.
@dusk125 @Elbehery We should point this out by checking this failure mode and see if etcd has any mechanisms to detect this. I suspect no, and it will eventually exhaust the disk space if the limit is greater than the available storage and not applicable but would be good to confirm.

Or actually if the db in usage size grows larger than the available memory, I would think the etcd pod could get OOM killed as well.

Reply:
etcd does not detect this, but we can add an alert. Not sure we have disk space metrics in prometheus though. 

Also imagine the case where one CP node has less disk available (eg due to bigger log files on the same partition). In an ideal world we would need to gather the metrics for all CP node disks utilization and decide.

Ideally would be a quota, or even a dedicated disk drive / partition that we can rely on - but that's sadly not the happy path installation experience in OCP as of today.

> db in usage size grows larger than the available memory

That's okay and not a problem. It's a memory mapped file, not an in-memory-file :)

Given the perf test results from [1] I would also assume apiserver going OOM long before etcd even gets there. Where we both have CP memory alerts on. 

[1] https://docs.google.com/presentation/d/1a_LiZzWgCepx0ywrO4nmNqMEpdtuU0j9CcrSViX6Vjs/edit#slide=id.g2a7e4538e11_0_64

Reply:
See my other comment about ram, which should also be cross-checked with storage.